rdfproxy/mapper.py

```diff
@@ -0,0 +1,87 @@
+"""ModelBindingsMapper: Functionality for mapping binding maps to a Pydantic model."""
+
+from collections.abc import Iterator
+from typing import Any, Generic, get_args
+
+from pydantic import BaseModel
+from rdfproxy.utils._types import _TModelInstance
+from rdfproxy.utils.utils import (
+    _collect_values_from_bindings,
+    _get_group_by,
+    _get_key_from_metadata,
+    _list_basemodel_p,
+    _list_p,
+)
+
+
+class ModelBindingsMapper(Generic[_TModelInstance]):
+    """Utility class for mapping flat bindings to a (potentially nested) Pydantic model."""
+
+    def __init__(self, model: type[_TModelInstance], *bindings: dict):
+        self.model = model
+        self.bindings = bindings
+        self._contexts = []
+
+    def get_models(self) -> list[_TModelInstance]:
+        """Generate a list of (potentially nested) Pydantic models based on (flat) bindings."""
+        return self._get_unique_models(self.model, self.bindings)
+
+    def _get_unique_models(self, model, bindings):
+        """Call the mapping logic and collect unique and non-empty models."""
+        models = []
+        for _bindings in bindings:
+            _model = model(**dict(self._generate_binding_pairs(model, **_bindings)))
+
+            if any(_model.model_dump().values()) and (_model not in models):
+                models.append(_model)
+
+        return models
+
+    def _get_group_by(self, model, kwargs) -> str:
+        """Get the group_by value from a model and register it in self._contexts."""
+        group_by: str = _get_group_by(model, kwargs)
+
+        if group_by not in self._contexts:
+            self._contexts.append(group_by)
+
+        return group_by
+
+    def _generate_binding_pairs(
+        self,
+        model: type[BaseModel],
+        **kwargs,
+    ) -> Iterator[tuple[str, Any]]:
+        """Generate an Iterator[tuple] projection of the bindings needed for model instantation."""
+        for k, v in model.model_fields.items():
+            if _list_basemodel_p(v.annotation):
```

Comment:
wouldn't that code below be a good candidate for `match`?

Reply:
The last two code paths (instance check and `else`) could be done concisely with match/case but checking against callables (`_list_basemodel_p` and `_list_p`) would require [guards](https://peps.python.org/pep-0622/#guards) here (I think).

Example:

```python
def some_predicate(value: str) -> bool:
    return True


if some_predicate("some value"):
    print("From conditional: Ok")
else:
    print("From conditional: Nope")

match value := "some value":
    case _ if some_predicate(value):
        print("From match/case: Ok")
    case _:
        print("From match/case: Nope")
```

I started out doing this with match/case but then again was reminded that match/case is not really a classical switch and more a pattern matching mechanism. So although it is possible to use match/case for this, I find if/elif/else more apt in this case.